subprojects/execution/src/main/java/org/gradle/internal/execution/impl/OutputsCleaner.java

```diff
@@ -0,0 +1,98 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.execution.impl;
+
+import com.google.common.collect.Ordering;
+import org.gradle.internal.file.FileType;
+import org.gradle.internal.fingerprint.FileCollectionFingerprint;
+import org.gradle.internal.fingerprint.FileSystemLocationFingerprint;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.util.Map;
+import java.util.PriorityQueue;
+import java.util.function.Predicate;
+
+public class OutputsCleaner {
+    private final PriorityQueue<File> directoriesToDelete;
+    private final Predicate<File> fileSafeToDelete;
+    private final Predicate<File> dirSafeToDelete;
+
+    private boolean didWork;
+
+    public OutputsCleaner(Predicate<File> fileSafeToDelete, Predicate<File> dirSafeToDelete) {
+        this.fileSafeToDelete = fileSafeToDelete;
+        this.dirSafeToDelete = dirSafeToDelete;
+        this.directoriesToDelete = new PriorityQueue<>(10, Ordering.natural().reverse());
+    }
+
+    public void cleanupOutputs(FileCollectionFingerprint fileCollectionFingerprint) {
+        for (Map.Entry<String, FileSystemLocationFingerprint> entry : fileCollectionFingerprint.getFingerprints().entrySet()) {
+            cleanupOutput(entry.getKey(), entry.getValue().getType());
+        }
+        cleanupDirectories();
+    }
+
+    private void cleanupOutput(String absolutePath, FileType fileType) {
+        switch (fileType) {
+            case Directory:
+                markDirForDeletion(new File(absolutePath));
+                break;
+            case RegularFile:
+                File file = new File(absolutePath);
+                if (fileSafeToDelete.test(file)) {
+                    didWork |= file.delete();
+                    markParentDirForDeletion(file);
+                }
+                break;
+            case Missing:
+                // Ignore missing files
+                break;
+            default:
+                throw new AssertionError("Unknown file type: " + fileType);
+        }
+    }
+
+    public boolean getDidWork() {
+        return didWork;
+    }
+
+    private void markParentDirForDeletion(File f) {
+        markDirForDeletion(f.getParentFile());
+    }
+
+    private void markDirForDeletion(@Nullable File dir) {
+        if (dir != null && dirSafeToDelete.test(dir)) {
+            directoriesToDelete.add(dir);
+        }
+    }
+
+    private void cleanupDirectories() {
+        while (!directoriesToDelete.isEmpty()) {
+            File directory = directoriesToDelete.poll();
+            if (isEmpty(directory)) {
+                didWork |= directory.delete();
+                markParentDirForDeletion(directory);
+            }
+        }
+    }
+
+    private boolean isEmpty(File parentDir) {
+        String[] children = parentDir.list();
```

Comment:
Seems like we should be using some kind of directory walking here instead. We could use the `FileSystemMirror`'s understanding of the file system state, too, no?

Reply:
We could use the full output snapshot if we wanted to. I think though that we should invest in making overlapping output cleanup much faster. It shouldn't happen too often anyway.

Reply:
> I think though that we should invest in making overlapping output cleanup much faster.

I'm confused. Should we invest in it or not?

Reply:
"not". I edited my comment.